rbac: add support for matching on route metadata

agrawroh opened the pull request:
## Background

This PR adds a new option called `sourced_metadata` to RBAC permissions and principals which allows specifying an optional source for the metadata to be matched. Currently it only supports Dynamic Metadata and Route Metadata. More options could be added later.

**Fixes:** #34913

---

**Commit Message:** rbac: add support for matching on route metadata
**Additional Description:** This PR adds a new option called `sourced_metadata` to RBAC permissions and principals which allows specifying an optional source for the metadata to be matched.
**Risk Level:** Low
**Testing:** Added Unit & Integration Tests
**Docs Changes:** Added
**Release Notes:** Added

agrawroh commented:
@yanavlasov @wbpcode Could you please help review this?

wbpcode commented on api/envoy/config/rbac/v3/rbac.proto:
```diff
@@ -366,6 +396,10 @@ message Principal {
     // ``not_id`` would match, this principal would not match. Conversely, if the
     // value of ``not_id`` would not match, this principal would match.
     Principal not_id = 8;
+
+    // Matches against metadata from either dynamic state or route configuration. Preferred over the
+    // ``metadata`` field as it provides more flexibility in metadata source selection.
+    SourcedMetadata sourced_metadata = 13;
```
I think you may need to deprecate the previsous `metadata` field. And I am not good at the naming, But the `SourcedMetadata` sounds a little weird.

agrawroh replied:
The intention was to convey that this would allow picking a `Source` for the metadata. Would `resolved_metadata` or `composite_metadata` be cleaner? If you have any other suggestions, I can change it to that.

wbpcode replied:
I also have no any idea. Compare to the new two name, seems sourced_metadata is fine. orz.

wbpcode commented on source/extensions/filters/common/rbac/matchers.cc:
```diff
@@ -248,7 +258,11 @@ bool AuthenticatedMatcher::matches(const Network::Connection& connection,
 bool MetadataMatcher::matches(const Network::Connection&, const Envoy::Http::RequestHeaderMap&,
                               const StreamInfo::StreamInfo& info) const {
-  return matcher_.match(info.dynamicMetadata());
+  if (metadata_source_ == envoy::config::rbac::v3::MetadataSource::DYNAMIC) {
+    return matcher_.match(info.dynamicMetadata());
+  }
+
+  return matcher_.match(info.route()->metadata());
```
How about the case where the route() is nullptr?

agrawroh replied:
That's a fair concern! I'll add a condition to handle that. Would matching on an empty metadata be appropriate in that case or should we fallback to matching on dynamic metadata? I personally think that we should match on empty metadata and maybe log a line in debug to convey that the route is missing.

wbpcode reviewed (commented):
Thanks for this contribution. I am fine to this new feature overall. Only some minor comments are added.

agrawroh commented:
> Thanks for this contribution. I am fine to this new feature overall. Only some minor comments are added.

@wbpcode Thank you so much for taking a look. I addressed your feedback and am happy to do the changes as needed.

wbpcode commented on api/envoy/config/rbac/v3/rbac.proto:
```diff
@@ -28,6 +28,14 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
+  // Query :ref:`dynamic metadata <well_known_dynamic_metadata>`
+  DYNAMIC = 0;
+
+  // Query :ref:`route entry metadata <envoy_v3_api_field_config.route.v3.Route.metadata>`
+  ROUTE_ENTRY = 1;
```
`ROUTE`, to keep consistance with our %METADATA%

agrawroh replied:
I was following the RateLimit use-case which is calling it [ROUTE_ENTRY](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#enum-config-route-v3-ratelimit-action-metadata-source). Should we keep this name or change it to `ROUTE`?

agrawroh commented:
/retest

agrawroh commented:
/retest

wbpcode commented on api/envoy/config/rbac/v3/rbac.proto:
```diff
@@ -274,12 +300,16 @@ message Permission {
     // URI template path matching.
     // [#extension-category: envoy.path.match]
     core.v3.TypedExtensionConfig uri_template = 13;
+
+    // Matches against metadata from either dynamic state or route configuration. Preferred over the
+    // ``metadata`` field as it provides more flexibility in metadata source selection.
+    SourcedMetadata sourced_metadata = 14;
```
Please mark and docmuent that the previous `metadata` is deprecated. In most cases, we will not keep two field with same featue. Thanks.

agrawroh replied:
Done, Thanks!

wbpcode commented on api/envoy/config/rbac/v3/rbac.proto:
```diff
@@ -193,8 +201,26 @@ message Policy {
+message SourcedMetadata {
+  // Metadata matcher configuration that defines what metadata to match against. This includes the filter name,
+  // metadata key path, and expected value.
+  type.matcher.v3.MetadataMatcher metadata_matcher = 1;
```
Please add a message vaidator to mark this field is required. Thanks.

agrawroh replied:
Done :)

wbpcode commented on test/extensions/filters/common/rbac/matchers_test.cc:
```diff
@@ -430,11 +430,13 @@ TEST(MetadataMatcher, MetadataMatcher) {
-  checkMatcher(MetadataMatcher(Matchers::MetadataMatcher(matcher, context)), false, conn, header,
-               info);
+  checkMatcher(MetadataMatcher(Matchers::MetadataMatcher(matcher, context),
+                               envoy::config::rbac::v3::MetadataSource::DYNAMIC),
+               false, conn, header, info);
   matcher.mutable_value()->mutable_string_match()->set_exact("prod");
-  checkMatcher(MetadataMatcher(Matchers::MetadataMatcher(matcher, context)), true, conn, header,
-               info);
+  checkMatcher(MetadataMatcher(Matchers::MetadataMatcher(matcher, context),
+                               envoy::config::rbac::v3::MetadataSource::DYNAMIC),
+               true, conn, header, info);
```
Please also add a `SourcedMetadata` matcher unit test.

agrawroh replied:
Added some Unit Tests

wbpcode commented:
Thanks for the update. Some new new minor comments are flushed.

cc @yanavlasov for a second pass for this rbac change. :)

agrawroh commented:
/retest

agrawroh commented:
/retest

agrawroh commented:
/retest

wbpcode commented on api/envoy/config/rbac/v3/rbac.proto:
```diff
@@ -237,8 +264,13 @@ message Permission {
+    type.matcher.v3.MetadataMatcher metadata = 7 [
+      deprecated = true,
+      (envoy.annotations.deprecated_at_minor_version) = "3.0",
+      (envoy.annotations.disallowed_by_default) = true
```
Thanks so much. But I think we doesn't need to add this  (envoy.annotations.disallowed_by_default) now.

wbpcode commented on api/envoy/config/rbac/v3/rbac.proto:
```diff
@@ -356,8 +392,13 @@ message Principal {
+    type.matcher.v3.MetadataMatcher metadata = 7 [
+      deprecated = true,
+      (envoy.annotations.deprecated_at_minor_version) = "3.0",
+      (envoy.annotations.disallowed_by_default) = true
```
ditto.

wbpcode commented on test/extensions/filters/common/rbac/matchers_test.cc:
```diff
@@ -624,6 +626,336 @@ TEST(UriTemplateMatcher, NoPathInHeader) {
+  }
+}
+
+TEST(PortRangeMatcher, EdgeCases) {
+  Envoy::Network::MockConnection conn;
+  Envoy::Http::TestRequestHeaderMapImpl headers;
+  NiceMock<StreamInfo::MockStreamInfo> info;
+
+  // Test boundary conditions
+  envoy::type::v3::Int32Range range;
+  range.set_start(1);
+  range.set_end(65535);
+
+  // Test port at start of range
+  Envoy::Network::Address::InstanceConstSharedPtr addr1 =
+      Envoy::Network::Utility::parseInternetAddressNoThrow("1.2.3.4", 1, false);
+  info.downstream_connection_info_provider_->setLocalAddress(addr1);
+  checkMatcher(PortRangeMatcher(range), true, conn, headers, info);
+
+  // Test port at end of range - 1
+  Envoy::Network::Address::InstanceConstSharedPtr addr2 =
```
Seems these are unreleated to this PR?

agrawroh replied:
Yeah, I was trying to increase the coverage of this class. I'll move it to a separate PR

wbpcode reviewed (commented):
LGTM overall. Only some nit comments to the annotation and tests. Then we are good to land it.

wbpcode commented:
/wait

agrawroh commented:
@wbpcode I'm trying to add an integration test which still uses `metadata` but am seeing these errors due to deprecation:
```
details: type envoy.config.rbac.v3.Permission Using deprecated option 'envoy.config.rbac.v3.Permission.metadata' from file rbac.proto. This configuration will be removed from Envoy soon. Please see https://www.envoyproxy.io/docs/envoy/latest/version_history/version_history for details. If continued use of this field is absolutely necessary, see https://www.envoyproxy.io/docs/envoy/latest/configuration/operations/runtime#using-runtime-overrides-for-deprecated-features for how to apply a temporary and highly discouraged override.
```

Any suggestions on how to fix this?

wbpcode commented:
> @wbpcode I'm trying to add an integration test which still uses `metadata` but am seeing these errors due to deprecation:
> 
> ```
> details: type envoy.config.rbac.v3.Permission Using deprecated option 'envoy.config.rbac.v3.Permission.metadata' from file rbac.proto. This configuration will be removed from Envoy soon. Please see https://www.envoyproxy.io/docs/envoy/latest/version_history/version_history for details. If continued use of this field is absolutely necessary, see https://www.envoyproxy.io/docs/envoy/latest/configuration/operations/runtime#using-runtime-overrides-for-deprecated-features for how to apply a temporary and highly discouraged override.
> ```
> 
> Any suggestions on how to fix this?

You can use `DEPRECATED_FEATURE_TEST` to wrap that test name.

agrawroh commented on api/envoy/config/rbac/v3/rbac.proto:
```diff
@@ -237,8 +264,12 @@ message Permission {
-    type.matcher.v3.MetadataMatcher metadata = 7;
+    // Metadata that describes additional information about the action. This field is deprecated; please use
+    // :ref:`sourced_metadata<envoy_v3_api_field_config.rbac.v3.Permission.sourced_metadata>` instead.
+    type.matcher.v3.MetadataMatcher metadata = 7 [
+      deprecated = true,
+      (envoy.annotations.deprecated_at_minor_version) = "3.0"
+    ];
```
```suggestion
    type.matcher.v3.MetadataMatcher metadata = 7
        [deprecated = true, (envoy.annotations.deprecated_at_minor_version) = "3.0"];
```

agrawroh commented on api/envoy/config/rbac/v3/rbac.proto:
```diff
@@ -356,8 +391,12 @@ message Principal {
-    type.matcher.v3.MetadataMatcher metadata = 7;
+    // Metadata that describes additional information about the principal. This field is deprecated; please use
+    // :ref:`sourced_metadata<envoy_v3_api_field_config.rbac.v3.Principal.sourced_metadata>` instead.
+    type.matcher.v3.MetadataMatcher metadata = 7 [
+      deprecated = true,
+      (envoy.annotations.deprecated_at_minor_version) = "3.0"
+    ];
```
```suggestion
    type.matcher.v3.MetadataMatcher metadata = 7
        [deprecated = true, (envoy.annotations.deprecated_at_minor_version) = "3.0"];
```

agrawroh commented:
@wbpcode It's ready for a final review and I'm working on the next PR to check in some more tests.

agrawroh commented:
@wbpcode It's ready for a final review and I'm working on the next PR to check in some more tests.

wbpcode commented:
LGTM. Thanks so much for this great changes.
